File: worker/src/rag_worker/embedding_provider.py
import hashlib
from logging import Logger

from openai import OpenAI

from rag_worker.config import WorkerSettings


class EmbeddingProvider:
    def __init__(self, settings: WorkerSettings, logger: Logger) -> None:
        self.settings = settings
        self.logger = logger
        self.client: OpenAI | None = None
        if settings.openai_api_key:
            self.client = OpenAI(api_key=settings.openai_api_key, timeout=settings.worker_openai_timeout_seconds)

    def _fallback_embedding(self, text: str) -> list[float]:
        # Deterministic fallback used when OPENAI_API_KEY is unavailable.
        digest = hashlib.sha256(text.encode("utf-8")).digest()
        vector = [0.0] * self.settings.worker_embedding_dim

        for idx in range(self.settings.worker_embedding_dim):
            byte = digest[idx % len(digest)]
            vector[idx] = (byte / 127.5) - 1.0

        return vector

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        if not texts:
            return []

        if not self.client:
            return [self._fallback_embedding(text) for text in texts]

        vectors: list[list[float]] = []
        batch_size = max(1, self.settings.worker_embedding_batch_size)

        for index in range(0, len(texts), batch_size):
            batch = texts[index : index + batch_size]
            response = self.client.embeddings.create(model=self.settings.worker_embedding_model, input=batch)
            vectors.extend([item.embedding for item in response.data])

        return vectors

File: lib/ingestion/runtime/embedding-provider.ts
import type {
  IngestionRuntimeSettings,
  RuntimeLogger,
} from "@/lib/ingestion/runtime/types";

type OpenAiEmbeddingResponse = {
  data?: Array<{
    embedding?: number[];
  }>;
  error?: {
    message?: string;
  };
};

export class EmbeddingProvider {
  private readonly apiKey: string | null;
  private readonly settings: IngestionRuntimeSettings;
  private readonly logger: RuntimeLogger;

  constructor(settings: IngestionRuntimeSettings, logger: RuntimeLogger) {
    this.settings = settings;
    this.logger = logger;
    this.apiKey = settings.openAiApiKey;
  }

  async embedTexts(texts: string[]): Promise<number[][]> {
    if (texts.length === 0) {
      return [];
    }

    if (!this.apiKey) {
      // Previously this silently wrote SHA-256-derived fake vectors that were
      // indistinguishable from real embeddings in the database. Failing the
      // job (retry -> dead letter) is the only honest behaviour: corrupt
      // embeddings poison retrieval for every future query.
      this.logger.error("embedding_api_key_missing", {
        model: this.settings.embeddingModel,
      });
      throw new Error(
        "OPENAI_API_KEY is not configured; refusing to ingest without real embeddings",
      );
    }

    const vectors: number[][] = [];
    const batchSize = Math.max(1, this.settings.embeddingBatchSize);

    for (let index = 0; index < texts.length; index += batchSize) {
      const batch = texts.slice(index, index + batchSize);
      const response = await fetch("https://api.openai.com/v1/embeddings", {
        method: "POST",
        signal: AbortSignal.timeout(this.settings.openAiTimeoutSeconds * 1000),
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model: this.settings.embeddingModel,
          input: batch,
          ...(this.settings.embeddingDimensions
            ? { dimensions: this.settings.embeddingDimensions }
            : {}),
        }),
      });

      const payload = (await response.json()) as OpenAiEmbeddingResponse;
      if (!response.ok) {
        const message =
          payload.error?.message ??
          `Embedding provider request failed (status=${response.status})`;
        this.logger.warn("embedding_request_failed", {
          message,
          batchSize: batch.length,
        });
        throw new Error(message);
      }

      for (const item of payload.data ?? []) {
        const embedding = item.embedding ?? [];
        vectors.push(embedding);
      }
    }

    return vectors;
  }
}
